Action-based arm control

kmdalal opened the pull request:
This branch replaces ROS Service clients with Action clients in the autonomy node.
It also removes the prior dummy GuardedMove action server and client.
This is an initial basic implementation that does nothing sophisticated with the Action messages.

It must be built and run with the corresponding branch in ow_simulator: 
  OCEANWATER-534_implement_current_arm_services_as_rosaction

To test it, start the simulator (any world) and try this plan to test all arm operations:
  roslaunch ow_autonomy autonomy_node.launch plan:=TestActions.plx

Also try the reference mission:
  roslaunch ow_autonomy autonomy_node.launch plan:=ReferenceMission1.plx

Note that the biggest code changes are in OwInterface.cpp.  There is also some unrelated refactoring of plan header files and minor code reformatting (mainly shortening long lines from others' commits).

anjan1984 commented:
I am getting this error 

process[autonomy_node-1]: started with pid [2137]
WARNING: AdapterFactory.cc:114: AdapterFactory: unable to load module for adapter type "CheckpointAdapter"
WARNING: AdapterConfiguration.cc:604: constructInterfaces: failed to construct adapter type "CheckpointAdapter"
[ERROR] [1617740825.036181285]: Interface initialization failed
[ERROR] [1617740825.036237377]: plexilInitializeInterfaces failed
[ERROR] [1617740825.036251543]: Could not initialize OW executive, shutting down.
[autonomy_node-1] process has died [pid 2137, exit code 1, cmd /usr/local/home/achakra2/oceanwaters_ws/devel/lib/ow_autonomy/autonomy_node ReferenceMission1.plx /ant_pan_position_controller/command:=/_original/ant_pan_position_controller/command /ant_tilt_position_controller/command:=/_original/ant_tilt_position_controller/command __name:=autonomy_node __log:=/home/achakra2/.ros/log/634e01d0-9716-11eb-8e74-104a7da983af/autonomy_node-1.log].
log file: /home/achakra2/.ros/log/634e01d0-9716-11eb-8e74-104a7da983af/autonomy_node-1*.log
all processes on machine have died, roslaunch will exit
shutting down processing monitor...
... shutting down processing monitor complete
done


-- am I missing something?

anjan1984 reviewed (approved):
I have no other comments other than the one I mentioned. The plans work fine for me.

Samahu commented on launch/autonomy_node.launch:
```diff
@@ -7,11 +7,9 @@
-        name="GuardedMoveServer"
-        type="GuardedMoveServer"
-        output="screen" />
+		<remap from="/ant_pan_position_controller/command"
+           to="/_original/ant_pan_position_controller/command"/>
+    <remap from="/ant_tilt_position_controller/command"
+           to="/_original/ant_tilt_position_controller/command"/>         
+  </node>
```
I don't understand why the autonomy node needs to do this re-mapping? The remapping is already done in `common.launch`. I don't remember when this was checked but I don't think this is right.

kmdalal replied:
None of this my code, and I don't recall doing this edit.  

The basic problem here, and in your previous comments, is that this branch has gotten out of date with OW-534 in ow_simulator.   I can see what the issues are.  I'll get on this shortly and let you know when it's ready.

Samahu reviewed (commented):
I got two issues while trying to run this branch `OCEANWATER-586-action-clients` along with `OCEANWATER-534_implement_current_arm_services_as_rosaction`

1)  This gets printed out right after executing `roslaunch ow_autonomy autonomy_node.launch plan:=ReferenceMission1.plx`

```bash
[ERROR] [1618609749.954996342, 1916524843.059999943]: Client [/autonomy_node] wants topic /power_system_node/remaining_useful_life to have datatype/md5sum [std_msgs/Float64/fdb28210bfa9d7c91146260178d9a584], but our version has [std_msgs/Int16/8524586e34fbd7cb1c08c5f5f1ca0e57]. Dropping connection.
```

2) Next problem happened when the autonomy node issued a guarded move command
```bash
[ERROR] [1618609844.867307, 1916524929.747500]: Exception in your execute callback: topic was closed during publish()
Traceback (most recent call last):
  File "/opt/ros/melodic/lib/python2.7/dist-packages/actionlib/simple_action_server.py", line 289, in executeLoop
    self.execute_callback(goal)
  File "/home/unaal/oceanwaters_ws/src/ow_simulator/ow_lander/scripts/unstow_action_server.py", line 84, in on_unstow_action
    self._update_feedback()
  File "/home/unaal/oceanwaters_ws/src/ow_simulator/ow_lander/scripts/unstow_action_server.py", line 49, in _update_feedback
    self._server.publish_feedback(self._fdbk)
  File "/opt/ros/melodic/lib/python2.7/dist-packages/actionlib/simple_action_server.py", line 175, in publish_feedback
    self.current_goal.publish_feedback(feedback)
  File "/opt/ros/melodic/lib/python2.7/dist-packages/actionlib/server_goal_handle.py", line 213, in publish_feedback
    self.action_server.publish_feedback(self.status_tracker.status, feedback)
  File "/opt/ros/melodic/lib/python2.7/dist-packages/actionlib/action_server.py", line 195, in publish_feedback
    self.feedback_pub.publish(af)
  File "/opt/ros/melodic/lib/python2.7/dist-packages/rospy/topics.py", line 882, in publish
    self.impl.publish(data)
  File "/opt/ros/melodic/lib/python2.7/dist-packages/rospy/topics.py", line 1101, in publish
    raise ROSException("topic was closed during publish()")
ROSException: topic was closed during publish()
```

kmdalal commented:
I merged melodic-devel, fixed conflicts, fixed an unrelated bug.  The previously reported errors should be gone.  However, TestActions, even though it runs to successful completion, does not reflect expected behavior in the simulator.  I have not tried these branches in weeks and something seems to have changed.  Reviewers, see how it goes for you...

Samahu reviewed (approved):
I did very limited testing to the autonomy part of the ros action interface since there were issues and behavior changes in the [main PR](https://github.com/nasa/ow_simulator/pull/66).

I won't be doing a code review this time for the reasons state in [my approve on the main PR](https://github.com/nasa/ow_simulator/pull/66#pullrequestreview-643517546)

kmdalal commented:
@Samahu,

> I did very limited testing to the autonomy part of the ros action interface since there were issues and behavior changes in the main PR.

The two tests of this PR (TestActions and ReferenceMission1 plan) should not encounter the behavior problems you found on the simulator branch, and any feedback would be helpful.

> I won't be doing a code review this time for the same reasons stated in my approve on the main PR

Despite your objection to the implementation strategy on the sim side, even a light code review would be helpful!  There's no one else with as much C++ expertise.

Samahu commented on src/plexil-adapter/OwInterface.h:
```diff
@@ -111,15 +134,16 @@ class OwInterface
-//////////////////// FAULTS FOR SYSTEM LEVEL ////////////////////////
-// structure of all maps for faults is the following:
-// key = (string) fault name
-// value = (pair) ( (int) numberical fault value, booleon of if fault exists)
+                          T msg_val,
+                          std::string key,
+                          T value,
+                          bool b );
```
Maybe it makes sense to pass string values as const ref since you are not modifying them inside the method

kmdalal replied:
Lannsie's addition, but I didn't catch that.  Changed.  However this function is deleted in my other branch!

Samahu reviewed (commented):
Sorry @mikedalal I'll try to do a more thorough approach on Monday

Samahu commented on src/plexil-adapter/OwInterface.cpp:
```diff
@@ -621,10 +577,32 @@ void OwInterface::initialize()
-    m_guardedMoveClient.reset(new GuardedMoveActionClient("GuardedMove", true));
-    m_guardedMoveClient->waitForServer();
-    ROS_INFO ("Action servers available.");
+    m_guardedMoveClient.reset(new GuardedMoveActionClient(Op_GuardedMove, true));
+    m_unstowClient.reset(new UnstowActionClient(Op_Unstow, true));
+    m_stowClient.reset(new StowActionClient(Op_Stow, true));
+    m_grindClient.reset(new GrindActionClient(Op_Grind, true));
+    m_digCircularClient.reset(new DigCircularActionClient(Op_DigCircular, true));
+    m_digLinearClient.reset(new DigLinearActionClient(Op_DigLinear, true));
+    m_deliverClient.reset(new DeliverActionClient(Op_Deliver, true));
```
since we have upgraded to C++14 you should be able to use `std::make_unique` which is more robust that doing reset(new ...) to set the a **unique_ptr**

kmdalal replied:
Created a more general and unassigned Jira OW-669 for this and also upgrading the raw pointers, which you can take if you like.  I think you were the one who introduced unique_ptr and reset() !

Samahu replied:
Yes, but back then we were still compiling against C++11 standard which lacked the `std::make_unique` function.

kmdalal replied:
Oh, I didn't know that.  I thought we were 14-compliant since the beginning of the project.

kmdalal commented on src/plexil-adapter/OwInterface.h:
```diff
@@ -87,12 +99,23 @@ class OwInterface
+                     double ground_pos, bool parallel, int id);
+  void digLinear1 (double x, double y, double depth, double length,
+                   double ground_pos, int id);
+  void deliver1 (double x, double y, double z, int id);
```
Good suggestion, done.

Samahu commented on src/plexil-adapter/OwInterface.cpp:
```diff
@@ -450,55 +421,44 @@ double OwInterface::groundPosition () const
 
-static void guarded_move_feedback_cb
-(const ow_autonomy::GuardedMoveFeedbackConstPtr& feedback)
+template<int OpIndex, typename T>
+static void action_done_cb
+(const actionlib::SimpleClientGoalState& state,
+ const T& result_ignored)
```
It doesn't seem like there is a need to templatize the second argument
```suggestion
template<int OpIndex>
static void action_done_cb
(const actionlib::SimpleClientGoalState& state,
 const ResultPtr& result_ignored)
```

kmdalal replied:
Your suggestion doesn't compile, as ResultPtr does not name a class.  I tried auto as well, not thinking that would work.

Samahu replied:
I didn't pay attention that these are specified as arguments to the template method and thought they were the same classes that were used. Please disregard this comment and related ones

Samahu commented on src/plexil-adapter/OwInterface.cpp:
```diff
@@ -450,55 +421,44 @@ double OwInterface::groundPosition () const
-static void guarded_move_done_cb
-(const actionlib::SimpleClientGoalState& state,
- const ow_autonomy::GuardedMoveResultConstPtr& result)
+template<typename T>
+static void action_feedback_cb (const T& feedback)
```
Again, it doesn't seem there is a purpose to the template here
```suggestion
static void action_feedback_cb (const FeedbackPtr& feedback)
```

kmdalal replied:
Your suggestion doesn't compile, as FeedbackPtr does not name a class.

Samahu commented on src/plexil-adapter/OwInterface.cpp:
```diff
@@ -600,9 +560,6 @@ void OwInterface::initialize()
     m_rulSubscriber = new ros::Subscriber
       (m_genericNodeHandle ->
        subscribe("/power_system_node/remaining_useful_life", qsize, rul_callback));
-    m_guardedMoveSubscriber = new ros::Subscriber
-      (m_genericNodeHandle ->
-       subscribe("/guarded_move_result", qsize, guarded_move_callback));
```
I don't see where this is set anywhere in the code... better remove it from the header and the destructor

kmdalal replied:
Good catch, it was a vestige of the old dummy action client example.

Samahu commented on src/plexil-adapter/OwInterface.cpp:
```diff
@@ -746,140 +646,235 @@ void OwInterface::takePicture (int id)
+  thread fault_thread (monitor_for_faults, opname);
+  if (ac) {
+    ac->sendGoal (goal,
+                  action_done_cb<OpIndex, ResultPtr>,
+                  active_cb<OpIndex>,
+                  action_feedback_cb<FeedbackPtr>);
```
Related to a note made earlier about the un-necessity of making the second arg of `action_done_cb` as a typename and the same for `action_feedback_cb`
```suggestion
                  action_done_cb<OpIndex>,
                  active_cb<OpIndex>,
                  action_feedback_cb);
```

kmdalal replied:
See comments above about the removal of templates not working.

Samahu commented on src/plexil-adapter/OwInterface.cpp:
```diff
@@ -746,140 +646,235 @@ void OwInterface::takePicture (int id)
-  if (check_service_client (client, Op_DeliverSample)) {
-    ow_lander::DeliverSample srv;
-    srv.request.x = x;
-    srv.request.y = y;
-    srv.request.z = z;
-    thread service_thread (call_ros_service<ow_lander::DeliverSample>,
-                           client, srv, Op_DeliverSample, id);
-    service_thread.detach();
+  if (m_guardedMoveClient) {
+    m_guardedMoveClient->sendGoal
+      (goal,
+       guarded_move_done_cb,
+       active_cb<GuardedMove>,
+       action_feedback_cb<ow_lander::GuardedMoveFeedbackConstPtr>);
   }
+  else {
+    ROS_ERROR ("m_guardedMoveClient was null!");
+    return;
+  }
+
+  // Wait for the action to return
+  bool finished_before_timeout =
+    m_guardedMoveClient->waitForResult (ros::Duration (ActionTimeoutSecs));
+
+  if (! finished_before_timeout) {
+    ROS_WARN ("GuardedMove action did not finish before the time out.");
+  }
+
+  mark_operation_finished (Op_GuardedMove, id);
+  fault_thread.join();
 }
```
You can avoid this code repetition/duplication simply by modifying the template method `runAction` adding a parameter to specify the callback you want to invoke with a default value of `action_done_cb<OpIndex>` passed to it.

kmdalal replied:
That looks doable, but I could not easily find out how to declare a template function parameter, and spent 30 minutes unsuccessfully on another approach: specializing action_done_cb for GuardedMove.   I can't spend more time on this, but feel free to implement your solution on a branch!

Samahu replied:
Please review #41

Samahu commented on src/plexil-adapter/README.md:
````diff
@@ -8,20 +8,3 @@ Contents
 environment), as well as the ROS autonomy node.
-
-
-ROS Action prototype
---------------------
-
-This directory also contains a dummy stand-in ROS action server for GuardedMove,
-as well as a standalone node for testing it.  (GuardedMoveServer,
-GuardedMoveClient).  Run the following sequence, in three separate shells.  Note
-that the latter two executables are found in the `devel/lib/ow_autonomy` directory
-of your OceanWATERS workspace.
-
-
-```bash
-  roscore
-  GuardedMoveServer
-  GuardedMoveClient
-  ```
````
You removed the old interfaces but shouldn't you put notes on the new ROS action interface?

kmdalal replied:
I removed the description of the dummy action prototype, but don't see a need to say anything about the ROS implementation or lander interface in this file -- these are documented in the user guide.

Samahu reviewed (commented):
I am leaving this review as a comment yet I think that m_guardedMoveSubscriber need to be removed since the it is no longer referenced or used. The other remarks should make the code better.

Samahu commented on src/plexil-adapter/OwInterface.cpp:
```diff
@@ -58,18 +48,35 @@ const double PanTiltTimeout = 5.0; // seconds, made up
 // In some cases, these must match those used in PLEXIL and/or ow_lander
-const string Op_GuardedMove       = "GuardedMove";
-const string Op_GuardedMoveAction = "GuardedMoveAction";
+const string Op_GuardedMove       = "Guarded_move";
```
Why does this use a different convention that the other commands?!

kmdalal replied:
Good catch, a vestige from when there was a duplicate GuardedMove.

kmdalal commented on src/plans/TestActions.plp:
```diff
@@ -2,35 +2,47 @@
+  
+  LibraryCall Unstow();
+
+  // Default args
+  //	LibraryCall GuardedMove (X=2, Y=0, Z=0.3, DirX=0, DirY=0, DirZ=1,
+  //                           SearchDistance=0.5);
```
Replaced with a comment indicating the default args, for convenient reference.

kmdalal commented on src/plans/TestActions.plp:
```diff
@@ -2,35 +2,47 @@
+	LibraryCall GuardedMove (X=1.75, Y=0.1, Z=0.2, DirX=0.1, DirY=0.1, DirZ=0.9,
+                           SearchDistance=0.7);
+
+  // Default args
+  //  LibraryCall Grind (X = 1.65, Y = 0, Depth = 0.05,
+  //                     Length = 0.6, Parallel = true,
+  //                     GroundPos = GP);
```
Same as previous.

kmdalal commented:
I've discovered a serious problem: every action will block autonomy for its timeout duration (currently 180 seconds) regardless of how fast it completes.  See calls to waitForResult.  I think this must be remedied.

kmdalal commented on src/plexil-adapter/OwInterface.cpp:
```diff
@@ -746,140 +646,235 @@ void OwInterface::takePicture (int id)
+
+  // Wait for the action to return
+  bool finished_before_timeout =
+    ac->waitForResult (ros::Duration (ActionTimeoutSecs));
```
This is a serious issue: every action will block autonomy for this timeout regardless of how fast it finishes.

Samahu replied:
Just use 0 for the duration, this should resolve it: 
```suggestion
    ac->waitForResult (ros::Duration (0));
```

kmdalal replied:
Will test with this...

kmdalal replied:
It worked fine, so I made this change.  Now, autonomy will wait indefinitely for actions to finish.  This is not good long-term, but the proper solution is out of scope for this ticket, and Release 8.

kmdalal commented on src/plexil-adapter/OwInterface.cpp:
```diff
@@ -746,140 +646,235 @@ void OwInterface::takePicture (int id)
+
+  // Wait for the action to return
+  bool finished_before_timeout =
+    m_guardedMoveClient->waitForResult (ros::Duration (ActionTimeoutSecs));
```
Serious problem: GuardedMove (and all actions) block for this timeout duration, regardless of how fast they complete.

kmdalal replied:
@Samahu suggests this may be a ROS issue.  waitForResult should not block for the timeout duration if the action finishes earlier.  And I have not consistently reproduced this problem.   Considering it out of scope of this review.

mogumbo commented on src/plans/TestActions.plp:
```diff
@@ -2,35 +2,44 @@
+
+  LibraryCall Unstow();
+
+  // Note: default args are (X=2, Y=0, Z=0.3, DirX=0, DirY=0, DirZ=1,
```
Why are the default args in these comments different from the ones in the function signatures? Wouldn't it be easier to make them the same and remove these comments?

kmdalal replied:
The intention was to use non-default arguments, but the comments are confusing; removed.

mogumbo commented on src/plans/plan-interface.h:
```diff
@@ -42,9 +42,15 @@ LibraryAction DigCircular (In Real X,
-LibraryAction DeliverSample (In Real X,
-                             In Real Y,
-                             In Real Z);
+LibraryAction DigLinear (In Real X,
```
Our coding guidelines say this should be named "digLinear". There are a lot of capitalized function names in here that should be lower-case.

lander-commands.h contains a corresponding `dig_linear` function. This is pretty confusing naming. Wouldn't it be better to give them more distinct names or put each of them inside a descriptive namespace? Or are these PLEXIL requirements that I don't understand?

kmdalal replied:
These particular examples follow the PLEXIL coding guidelines in our Confluence.  We can discuss if your interpretation of that section is different.

anjan1984 reviewed (approved):
The Reference mission completed successfully !!
